### PR TITLE
multi: single miner endpoint.

### PR DESCRIPTION
**Pull request by dnldd**

### CODE REVIEW

**Comment by dnldd** [comment]
Tested miner monitoring and upgrade logic with a `DR5` on `mainnet`. Initially recognized as a `DR3` and later upgrading to a `DR5` per the upgrade path.

```sh
2020-11-27 19:26:11.637 [INF] POOL: listening on :5550
2020-11-27 19:29:25.140 [DBG] POOL: Mining client connected. extranonce1=ca750c60, addr=166.170.24.8:35258
2020-11-27 19:29:32.007 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:29:33.659 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:29:40.999 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:29:46.569 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:29:48.109 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:29:52.109 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:29:56.349 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:30:06.370 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:30:09.189 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:30:15.849 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:30:22.246 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:30:25.340 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:30:38.170 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:30:46.229 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:30:50.579 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:31:01.999 [ERR] POOL: submitted work from ca750c60/antminerdr3 is not less than the network target difficulty
2020-11-27 19:31:25.143 [INF] POOL: upgrading ca750c60/antminerdr3 to ca750c60/antminerdr5
2020-11-27 19:31:25.143 [INF] POOL: updated difficulty (228174.031) for ca750c60/antminerdr5 sent
2020-11-27 19:33:06.719 [ERR] POOL: submitted work from ca750c60/antminerdr5 is not less than the network target difficulty
2020-11-27 19:34:43.039 [ERR] POOL: submitted work from ca750c60/antminerdr5 is not less than the network target difficulty
2020-11-27 19:35:13.256 [ERR] POOL: submitted work from ca750c60/antminerdr5 is not less than the network target difficulty
2020-11-27 19:35:38.966 [ERR] POOL: submitted work from ca750c60/antminerdr5 is not less than the network target difficulty
2020-11-27 19:35:46.516 [ERR] POOL: submitted work from ca750c60/antminerdr5 is not less than the network target difficulty
```

File: pool/endpoint.go [hunk 1]
```diff
@@ -14,8 +14,72 @@ import (
 	"time"
 
 	"github.com/decred/dcrd/chaincfg/v3"
+	errs "github.com/decred/dcrpool/errors"
 )
 
+var (
```
Comment (jholdstock) [inline pool/endpoint.go:20]: I think all of the miner ID related stuff is good candidate for a new file - potentially `miner_id.go`? Miner ID is a distinct concept to the endpoint concept, so arguably does not belong in `endpoint.go`.

- The ID strings
- `minerIDPair`
- `newMinerIDPair`
- `generateMinerIDs`
- `identifyMiner`

File: pool/endpoint.go [hunk 2]
```diff
@@ -51,6 +118,12 @@ type EndpointConfig struct {
 	FetchHostConnections func(string) uint32
 	// SignalCache sends the provided cache update event to the gui cache.
 	SignalCache func(event CacheUpdateEvent)
+	// MonitorCycle represents the time monitoring a mining client to access
+	// possible upgrades if needed
```
Comment (jholdstock) [inline pool/endpoint.go:122]:
```suggestion
	// possible upgrades if needed.
```

File: pool/hub.go [hunk 3]
```diff
@@ -120,22 +120,45 @@ type NodeConnection interface {
...
 	MaxConnectionsPerHost uint32
-	WalletAccount         uint32
-	CoinbaseConfTimeout   time.Duration
+	// WalletAccount represents the wallet account to process payments from.
+	WalletAccount uint32
+	// CoinbaseConfTimeout is the duration to wait for coinbase confirmations
+	// when generating a payout transaction.
+	CoinbaseConfTimeout time.Duration
+	// MonitorCycle represents the time monitoring a mining client to access
+	// possible upgrades if needed
```
Comment (jholdstock) [inline pool/hub.go:157]:
```suggestion
	// possible upgrades if needed.
```

**Review by jholdstock — CHANGES_REQUESTED** [review]

File: gui/assets/templates/index.html [hunk 4]
```diff
@@ -85,29 +86,11 @@ <h2>Identify the Miner</h2>
...
-                                    <li>
-                                        <span class="config">{{.MinerPorts.antminerdr5}}</span> (Antminer&nbsp;DR5)
-                                    </li>
-                                    <li>
-                                        <span class="config">{{.MinerPorts.whatsminerd1}}</span> (Whatsminer&nbsp;D1)
-                                    </li>
-                                </ul>
-                            </div>
-                            <p>The username of the miner should be the payment address for receiving rewards and a unique name identifying the client, formatted as <span class="config">'address.name'</span></p>
+                            <p>To connect, set the pool URL to <span class="config">'localhost'</span>, and the miner
```
[the remark this answers is not in the document]
Comment (dnldd) [reply gui/assets/templates/index.html:89]: Sure thing, let's do that.

File: gui/assets/templates/index.html [hunk 5]
```diff
@@ -72,7 +72,8 @@ <h2>Identify the Miner</h2>
                                     <li>Whatsminer&nbsp;D1</li>
                                 </ul>
                             </div>
-                            <p>For Innosilicon D9 the pool only supports <a href="https://github.com/decred/dcrpool/releases/download/v1.1.0-rc1/D9_20180602_094459.swu">D9_20180602_094459.swu</a> 
+                            <p>For Innosilicon D9 the pool only supports <a
```
[the remark this answers is not in the document]
Comment (dnldd) [reply gui/assets/templates/index.html:75]: Yes, will update the listing the upcoming commit.

File: pool/client.go [hunk 6]
```diff
@@ -249,15 +266,89 @@ func (c *Client) handleAuthorizeRequest(req *Request, allowed bool) error {
...
+			miner := pair.miners[idx]
+			newID := fmt.Sprintf("%v/%v", c.extraNonce1, miner)
+			log.Infof("upgrading %s to %s", c.id, newID)
+			c.miner = miner
+			c.id = newID
+			info, err := c.cfg.FetchMinerDifficulty(miner)
+			if err != nil {
+				tries++
+				log.Error(err)
+				continue
```
[the remark this answers is not in the document]
Comment (dnldd) [reply pool/client.go:335]: Good catch, will call `FetchMinerDifficulty` to ensure it doesn't error before setting `c.miner` and `c.id`.

File: pool/client.go [hunk 7]
```diff
@@ -269,21 +360,45 @@ func (c *Client) handleSubscribeRequest(req *Request, allowed bool) error {
...
+	}
+
+	c.mtx.Lock()
+	minerIdx := 0
+	miner := idPair.miners[minerIdx]
+	c.miner = miner
+	c.id = fmt.Sprintf("%v/%v", c.extraNonce1, miner)
+	info, err := c.cfg.FetchMinerDifficulty(miner)
+	if err != nil {
+		c.mtx.Unlock()
```
[the remark this answers is not in the document]
Comment (dnldd) [reply pool/client.go:387]: same here, will call `FetchMinerDifficulty` to ensure it doesn't error before setting `c.miner` and `c.id`.

File: pool/client.go [hunk 8]
```diff
@@ -249,15 +266,89 @@ func (c *Client) handleAuthorizeRequest(req *Request, allowed bool) error {
...
+				continue
+			}
+			c.diffInfo = info
+			c.mtx.Unlock()
+
+			c.setDifficulty()
+			log.Infof("updated difficulty (%s) for %s sent",
+				c.diffInfo.difficulty.FloatString(3), c.id)
+			time.Sleep(time.Millisecond * 500)
+			c.updateWork()
```
[the remark this answers is not in the document]
Comment (dnldd) [reply pool/client.go:344]: Agreed, makes sense to reset tries after an upgrade but subs doesn't need a reset, it's always the difference of the current sub count and the previous count therefore will always be greater than or equal to zero and reflect the number of subs within the period being measured.

File: gui/assets/templates/index.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (jholdstock) [reply gui/assets/templates/index.html:95]: #292

**Review by jholdstock — APPROVED** [review]
Looking good to me! Thanks for taking the time to split up the tests, they will definitely be more manageable going forward.

**Pull request merged**